registry/quilt_server/views.py

```diff
@@ -778,6 +787,47 @@ def package_put(owner, package_name, package_hash):
                     dict(team=app.config['TEAM_ID'], user=owner, pkg=package_name)
                 )
 
+    if package_path is not None:
+        # We're doing a subpackage push - so look up the existing contents.
+        result = (
+            db.session.query(Instance, Tag)
+            .filter_by(package=package)
+            .options(undefer('contents'))  # Contents is deferred by default.
+            .join(Instance.tags)
+            .filter_by(tag=LATEST_TAG)
+            .with_for_update()
+            .one_or_none()
+        )
+        if result is None:
+            raise ApiException(
+                requests.codes.not_found,
+                "Tag %r does not exist" % LATEST_TAG
+            )
+        base_instance, tag = result
+
+        subnode = contents
+        base_subnode = base_instance.contents
+        component = None
+        for component in package_path.split('/'):
+            try:
+                subnode = subnode.children[component]
+            except (AttributeError, KeyError):
+                raise ApiException(requests.codes.not_found, "Invalid source subpath: %r" % component)
+            base_subnode_parent = base_subnode
+            try:
+                base_subnode = base_subnode.children.get(component)
+            except AttributeError:
+                raise ApiException(requests.codes.not_found, "Target subpath is not a group node: %r" % component)
+            if base_subnode is None:
+                break
+        try:
+            base_subnode_parent.children[component] = subnode
```

Comment:
I can live with this limitation, but it's not ideal. It would be nice to support e.g., changing a single data node into group node with the original data in the folder.

Reply:
agree